components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,599 @@
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                // Get primary keystore from keyStoreManager
+                return keyStoreManager.getPrimaryKeyStore();
+            }
+
+            // Get tenant keystore from keyStoreManager
+            String tenantKeyStoreName = IdentityKeyStoreResolverUtil.buildTenantKeyStoreName(tenantDomain);
+            return keyStoreManager.getKeyStore(tenantKeyStoreName);
+        } catch (Exception e) {
+            throw new IdentityKeyStoreResolverException(
+                    ErrorMessages.ERROR_CODE_ERROR_RETRIEVING_TENANT_KEYSTORE.getCode(),
+                    String.format(ErrorMessages.ERROR_CODE_ERROR_RETRIEVING_TENANT_KEYSTORE.getDescription(),
+                            tenantDomain), e);
+        }
+    }
+
+    /**
+     * Return Primary, tenant or custom keystore.
+     *
+     * @param tenantDomain      Tenant domain
+     * @param inboundProtocol   Inbound authentication protocol of the application
+     * @return Primary, tenant or custom keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public KeyStore getKeyStore(String tenantDomain, InboundProtocol inboundProtocol)
+            throws IdentityKeyStoreResolverException {
+
+        if (tenantDomain == null || tenantDomain.isEmpty()) {
+            throw new IdentityKeyStoreResolverException(
+                    ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getCode(),
+                    String.format(ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getDescription(), "Tenant domain"));
+        }
+        if (inboundProtocol == null) {
+            throw new IdentityKeyStoreResolverException(
+                    ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getCode(),
+                    String.format(ErrorMessages.ERROR_CODE_INVALID_ARGUMENT.getDescription(), "Inbound protocol"));
+        }
+
+        if (keyStoreMappings.containsKey(inboundProtocol)) {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain) ||
+                    keyStoreMappings.get(inboundProtocol).getUseInAllTenants()) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Custom keystore configuration avialble for " + inboundProtocol + " protocol.");
+                }
+
+                String keyStoreName = IdentityKeyStoreResolverUtil.buildCustomKeyStoreName(
+                        keyStoreMappings.get(inboundProtocol).getKeyStoreName());
+
+                try {
+                    int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                    KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+                    return keyStoreManager.getKeyStore(keyStoreName);
```

Comment:
Shall we add a debug log when loading a keystore at either kernel level or framework level? Something like "Loading keystore for tenant: "

Reply:
Added the debug logs to KeyStoreManager in the kernel since it would benefit other flows as well
https://github.com/wso2/carbon-kernel/pull/4025/commits/bb1cd7d4bedaca490f87c5d25c805a6566207edc